### ISSUE TITLE
Ability to get segment results from a MiningModel with modelChain / selectFirst / ...

### ISSUE THREAD

**Issue by daniel-boehme** [jpmml/jpmml-evaluator#273]
I use `jpmml-evaluator` to understand where different PMML implementations differ in their results. This works very well for models like random forests (single `MiningModel` containing multiple `TreeModel`s), as the result implements `HasSegmentResults` to get the results of the individual trees, which implement `HasDecisionPath`.

However, I'm struggling to get similar insights from boosted trees (e.g. generated using `xgboost`). I think I know why, but I'm unsure if my proposed solution would even make sense.

The outline of the model looks like this (details omitted for brevity):
```
<MiningModel>
    <Segmentation multipleModelMethod="modelChain">
        <Segment id="1">
            <MiningModel>
                <OutputField/>
                <Segmentation multipleModelMethod="sum">
                    <TreeModel/>
                    <TreeModel/>
                    <TreeModel/>
                </Segmentation>
            </MiningModel>
        </Segment>
        <Segment id="2">
            <RegressionModel>
                <OutputField/>
            </RegressionModel>
        </Segment>
    </Segmentation>
</MiningModel>
```

From my (very basic and shallow) understanding of the code, the result of a `MiningModel` with `multipleModelMethod="sum"` is explicitly created to have segment results [here](https://github.com/jpmml/jpmml-evaluator/blob/f7e773e7f0790fedf06403ca79f5b48fb864b57f/pmml-evaluator/src/main/java/org/jpmml/evaluator/mining/MiningModelEvaluator.java#L279) (similar code exists in other functions in this class).

However, for `multipleModelMethod="modelChain"`, this is not the case, as the code seems to just return the result of the last segment [here](https://github.com/jpmml/jpmml-evaluator/blob/f7e773e7f0790fedf06403ca79f5b48fb864b57f/pmml-evaluator/src/main/java/org/jpmml/evaluator/mining/MiningModelEvaluator.java#L253).

So my question is:
Would it make sense to change `‎MiningModelEvaluator::selectFirst()`, `‎MiningModelEvaluator::selectAll()`, `‎MiningModelEvaluator::modelChain()`, and `MiningModelEvaluator::multiModelChain()` to return a `SegmentResult`, that also implements `HasSegmentResults` and returns the list of all the segment results?

Example:
```
	private Map<String, ?> modelChain(List<SegmentResult> segmentResults){
		SegmentResult result = segmentResults.get(segmentResults.size() - 1);
		return new SegmentResult(result.getSegment(), result.getResults()) {
			@Override
			public Collection<? extends SegmentResult> getSegmentResults(){
				return segmentResults;
			}
		};
	}
```

**Comment by vruusmann** [comment]
Noted.

From what I can remember "natively", the `HasSegmentResult` marker interface is designed for situations where you want to "step inside" the results of a Model (eg. analyzing what was aggregated). 

The situation appears a bit different with non-aggregating models, such as model chains. Here you're not "stepping inside", but you're interestid in "looking around at the same level". In other words, instead of accessing the results of "child models", you want to access the results of "sibling models"?

Creating a new instance of `SegmentResult` as in the last code example seems a bit misleading. It either needs to be instantiated at a different level (ie. parent), or use a different marker interface. For example, something like `HasSiblingSegmentResults`.

**Comment by vruusmann** [comment]
Also, multiple model methods (MMMs) 1) `selectFirst` and `selectAll` versus 2) `modelChain` and `multiModelChain` appear functionally different. The former are more aggregation-like (they are simply forming a collection, rather than calculating statistics as is the case with numerical aggregators such as `sum`, `weightedSum` etc), whereas the latter do not perform any aggregation at all.

It might be necessary to think deeper about the taxonomy of MMMs, and come up with multiple marker interfaces (possibly collected under a single super-marker interface).

Anyway, accessing the internals of a PMML prediction is a worthwhile cause (and, AFAIK, not possible with any other PMML engine other than JPMML-Evaluator), and I'll look deeper into it in the next version.

**Comment by daniel-boehme** [comment]
> Anyway, accessing the internals of a PMML prediction is a worthwhile cause (and, AFAIK, not possible with any other PMML engine other than JPMML-Evaluator), and I'll look deeper into it in the next version.

Thank you, I appreciate that.

I guess one could also argue that there is the result of the mining model, which is derived in _some_ form from multiple segment results. And this is completely independent from the derivation method (`sum`, `selectFirst`, `modelChain`, ...).
So at least when looking at it that way, it could make sense for any result of a mining model to implement `HasSegmentResults`. There is no distinction between same level or different level.

IMHO, it boils down to what is the desired abstraction level for the result of a mining model. Does it "hide" the multiple model method (in terms of API exposure), or not.

**Comment by vruusmann** [comment]
>  It boils down to what is the desired abstraction level for the result of a mining model.

The `MiningModel` element represent a meta-model, whereas all other model elements represent concrete/evaluatable models. To make things worse, its behaviour depends on the multiple model method (MMM) attribute. Perhaps the designers of the PMML specification should have separated normal/aggregation MMMs (eg. `sum`) and control flow/non-aggregation MMMs (eg. `modelChain`) into two separate model elements.

The current implementation of `modelChain` effectively propagates the result that it is given; it does not compute anything new by itself (or transform the incoming result).

For example, an XGBoost model is typically a two-element model chain, where the first element is a tree ensemble regressor, and the second element  is a (binary logistic-) classifier. Durng evaluation, this classifier yields an instance of `org.jpmml.evaluator.Classification`, and the to-level `MiningModel` element simply propagates it further.

This naive-ish result propagation behaviour doesn't lend itself too well for extension. For example, it's impossble to "dynamically attach" any new marker interfaces (such as `HasSegmentResult`) to an existing `o.j.e.Classification` object. One would have to instantiate a completely new classification subclass, and copy all the incoming classification state over there.

**Comment by vruusmann** [comment]
I was thinking about an alternative design.

All JPMML-Evaluator result classes (eg. `o.j.e.Classification`, `o.j.e.Regression`, etc) inherit from the `org.jpmml.evaluator.AbstractComputable` abstract base class. Perhaps this class should have a special field for holding "mining model context" information.

If a PMML document contains a top-level `RegressionModel` element, then this field would be left unset (ie. `null`). However, if a `RegressionModel` element (or any other model element) is nested inside a `MiningModel` element, then during each "result propagation", the `MiningModelEvaluator` would append a reference to itself into the "mining model context" chain.

If you want to debug a prediction, then you could traverse the list of "mining model contexts", and access their `HasSegmentResult` infos and other stuff.

TLDR: If a `o.j.e.Computable` object is being propagated through a `MiningModel` element, then it should accumulate something akin to Java's stack trace (eg. "this result was propagated through three levels of `MiningModel` elements, and at each level the evaluation end state was such-and-such"). Right now it accumulates nothing.

**Comment by vruusmann** [comment]
The return type of the `MiningModelEvaluator#evaluate(Map<String, ?>)` method is [`org.jpmml.evaluator.mining.SegmentationResult`](https://github.com/jpmml/jpmml-evaluator/blob/1.6.7/pmml-evaluator/src/main/java/org/jpmml/evaluator/mining/SegmentationResult.java) now (irrespective of the multiple model method - whether `sum`, `modelChain`, `x-multiModelChain`, or something else).

This class provides the `#getResults(Iterable<String>)` method, which lets you directly access any "internal" result map. The argument to this method is an iterable of `Segment` identifiers, designating the "mining path" from the top-level to the requested level.

Let's see this class in action based on your original PMML example. Please note that I tweaked it a bit so that every `Segment` element has an explicit `id` attribute now. 

```java
import org.jpmml.evaluator.mining.SegmentationResult

Evaluator evaluator = ...

// This cast is only valid for `MiningModelEvaluator` results
SegmentationResult xgbResults = (SegmentationResult)evaluator.evaluate(...)

// The regressor component is a `MiningModel` element, so its results are also `org.jpmml.evaluator.mining.SegmentationResult` objects!
SegmentationResult regressorResults = (SegmentationResult)xgbResults.getResults(Arrays.asList("step_1"));
Collection<? extends SegmentResult> regressorSegmentResults = regressorResults.getSegmentResults();
for(SegmentResult regressorSegmentResult : regressorSegmentResults){
	System.out.println(regressorSegmentResult);
}

// The classifier component is a `RegressionModel` element, so its results are plain `java.util.Map` objects
Map<String, ?> classifierResults = xgbResults.getResults(Arrays.asList("step_2"));

// Access specific `TreeModel` element results by specifing its "mining path":
Map<String, ?> tree3Results = xgbResults.getResults(Arrays.asList("step_1", "tree_3"));
```

**Comment by daniel-boehme** [comment]
Thank you very much, I appreciate your work.

And the way I understand it, I can use `SegmentationResult#getResults()` (without an argument) to just get all of them. And that's exactly what I need. I'll try it out the next days.

**Comment by daniel-boehme** [comment]
Update: I've tried it out and it works like charm. Thank you.
